src/NuGet.Core/NuGet.Packaging/Signing/Native/NativeUtility.cs

```diff
@@ -0,0 +1,178 @@
+                        }
+
+                        signedInfo.rgCertEncoded = certificatesHandle.DangerousGetHandle();
+
+                        var hMsg = NativeMethods.CryptMsgOpenToEncode(
+                            NativeMethods.X509_ASN_ENCODING | NativeMethods.PKCS_7_ASN_ENCODING,
+                            dwFlags: 0,
+                            dwMsgType: NativeMethods.CMSG_SIGNED,
+                            pvMsgEncodeInfo: ref signedInfo,
+                            pszInnerContentObjID: null,
+                            pStreamInfo: IntPtr.Zero);
+
+                        ThrowIfFailed(!hMsg.IsInvalid);
+
+                        ThrowIfFailed(NativeMethods.CryptMsgUpdate(
+                            hMsg,
+                            data,
+                            (uint)data.Length,
+                            fFinal: true));
+
+                        uint valueLength = 0;
+
+                        ThrowIfFailed(NativeMethods.CryptMsgGetParam(
+                            hMsg,
+                            CMSG_GETPARAM_TYPE.CMSG_CONTENT_PARAM,
+                            dwIndex: 0,
+                            pvData: null,
+                            pcbData: ref valueLength));
+
+                        encodedData = new byte[(int)valueLength];
+
+                        ThrowIfFailed(NativeMethods.CryptMsgGetParam(
+                            hMsg,
+                            CMSG_GETPARAM_TYPE.CMSG_CONTENT_PARAM,
+                            dwIndex: 0,
+                            pvData: encodedData,
+                            pcbData: ref valueLength));
+                    }
+                }
+
+                var cms = new SignedCms();
+
+                cms.Decode(encodedData);
+
+                return cms;
+            }
+        }
+
+        private unsafe static CMSG_SIGNER_ENCODE_INFO CreateEncodeInfo(
+            X509Certificate2 certificate,
+            CngKey privateKey,
+            CryptographicAttributeObjectCollection attributes,
+            Common.HashAlgorithmName hashAlgorithm,
+            HeapBlockRetainer hb)
+        {
+            var signerInfo = new CMSG_SIGNER_ENCODE_INFO();
+            signerInfo.cbSize = (uint)Marshal.SizeOf(signerInfo);
+            signerInfo.pCertInfo = Marshal.PtrToStructure<CERT_CONTEXT>(certificate.Handle).pCertInfo;
+            signerInfo.hCryptProvOrhNCryptKey = privateKey.Handle.DangerousGetHandle();
+            // DwKeySpec is not used when hCryptProvOrhNCryptKey is used.
+            signerInfo.dwKeySpec = 1;
+            signerInfo.HashAlgorithm.pszObjId = hashAlgorithm.ConvertToOidString();
+
+            if (attributes.Count != 0)
+            {
+                signerInfo.cAuthAttr = attributes.Count;
+
+                checked
+                {
+                    var attributeSize = Marshal.SizeOf(typeof(CRYPT_ATTRIBUTE));
```

Comment:
Can't you use `Marshal.SizeOf<T>()`?  I think there's a compiler warning for using the old `Marshal.SizeOf(Type)`.